File: mos/frontend/src/types/commerce.ts
import type { Product, ProductVariant } from "@/types/products";

// =============================================================================
// Legacy Funnel Commerce Types
// =============================================================================

export type PublicCommerceVariant = Omit<ProductVariant, "external_price_id">;

export type PublicCommerceProduct = Product & {
  variants: PublicCommerceVariant[];
  variants_count: number;
};

export type PublicFunnelCommerce = {
  productSlug: string;
  funnelSlug: string;
  funnelId: string;
  product: PublicCommerceProduct;
};

// =============================================================================
// Medusa Store API Response Types
// =============================================================================

export type MedusaRegion = {
  id: string;
  name: string;
  currency_code: string;
  tax_rate?: number;
  countries?: Array<{
    iso_2: string;
    display_name: string;
  }>;
};

export type MedusaProductVariant = {
  id: string;
  title: string;
  sku?: string;
  barcode?: string;
  allow_backorder?: boolean;
  manage_inventory?: boolean;
  inventory_quantity?: number;
  prices: Array<{
    id: string;
    currency_code: string;
    amount: number;
    original_amount?: number;
  }>;
  options?: Record<string, string>;
};

export type MedusaProduct = {
  id: string;
  title: string;
  handle: string;
  description?: string;
  thumbnail?: string;
  status?: string;
  variants?: MedusaProductVariant[];
  options?: Array<{
    id: string;
    title: string;
    values?: Array<{
      id: string;
      value: string;
    }>;
  }>;
  collection_id?: string;
  categories?: Array<{
    id: string;
    name: string;
  }>;
};

export type MedusaCollection = {
  id: string;
  title: string;
  handle: string;
  products?: MedusaProduct[];
};

export type MedusaCategory = {
  id: string;
  name: string;
  handle: string;
  description?: string;
  parent_category_id?: string;
  category_children?: Array<{
    id: string;
    name: string;
  }>;
};

export type MedusaCartAddress = {
  first_name?: string;
  last_name?: string;
  address_1?: string;
  address_2?: string;
  city?: string;
  province?: string;
  postal_code?: string;
  country_code?: string;
  phone?: string;
};

export type MedusaCartLineItem = {
  id: string;
  cart_id: string;
  title: string;
  description?: string;
  variant_id: string;
  quantity: number;
  unit_price: number;
  subtotal?: number;
  tax_total?: number;
  total?: number;
  variant?: MedusaProductVariant;
};

export type MedusaCart = {
  id: string;
  email?: string;
  region_id: string;
  shipping_address?: MedusaCartAddress;
  billing_address?: MedusaCartAddress;
  items?: MedusaCartLineItem[];
  shipping_methods?: Array<{
    id: string;
    shipping_option_id: string;
    price: number;
  }>;
  subtotal?: number;
  tax_total?: number;
  shipping_total?: number;
  discount_total?: number;
  total?: number;
  currency_code: string;
};

export type MedusaShippingOption = {
  id: string;
  name: string;
  price_type: string;
  amount?: number;
  currency_code?: string;
  region_id: string;
};

export type MedusaPaymentProvider = {
  id: string;
};

export type MedusaPaymentSession = {
  id: string;
  provider_id: string;
  status: string;
  amount: number;
  data?: Record<string, unknown>;
};

export type MedusaPaymentCollection = {
  id: string;
  status: string;
  amount: number;
  currency_code: string;
  payment_sessions?: MedusaPaymentSession[];
};

// =============================================================================
// Site Commerce API Types
// =============================================================================

export type SiteCommerceData = {
  // Site metadata
  siteFamily?: string;
  commerceProvider?: string;
  storeName?: string;  // Store name for branding

  // Catalog data
  regions: MedusaRegion[];
  products: MedusaProduct[];
  collections: MedusaCollection[];
  categories: MedusaCategory[];

  // Current product (if specified)
  currentProduct?: MedusaProduct;

  // Current category (if category_handle provided)
  currentCategory?: MedusaCategory;

  // Cart data (if cart_id provided)
  cart?: MedusaCart;

  // Shipping options (if cart_id provided)
  shippingOptions?: MedusaShippingOption[];

  // Payment providers (if region_id provided)
  paymentProviders?: MedusaPaymentProvider[];

  // Pagination
  productsCount?: number;

  // Errors from critical fetches
  errors?: string[];
};

export type SiteCommerceCartResponse = {
  cart: MedusaCart;
};

export type SiteCommerceShippingOptionsResponse = {
  shipping_options: MedusaShippingOption[];
};

export type SiteCommercePaymentProvidersResponse = {
  payment_providers: MedusaPaymentProvider[];
};

export type SiteCommercePaymentSessionResponse = {
  payment_collection: MedusaPaymentCollection;
};

export type SiteCommerceCompleteResponse = {
  type: "order" | "cart";
  order?: Record<string, unknown>;
  cart?: MedusaCart;
};

// =============================================================================
// Cart State Management Types
// =============================================================================

export type CartState = {
  cartId: string | null;
  cart: MedusaCart | null;
  loading: boolean;
  error: string | null;
};

export type CartAction =
  | { type: "SET_CART_ID"; payload: string | null }
  | { type: "SET_CART"; payload: MedusaCart | null }
  | { type: "SET_LOADING"; payload: boolean }
  | { type: "SET_ERROR"; payload: string | null };

export const cartReducer = (state: CartState, action: CartAction): CartState => {
  switch (action.type) {
    case "SET_CART_ID":
      return { ...state, cartId: action.payload };
    case "SET_CART":
      return { ...state, cart: action.payload };
    case "SET_LOADING":
      return { ...state, loading: action.payload };
    case "SET_ERROR":
      return { ...state, error: action.payload };
    default:
      return state;
  }
};

export const initialCartState: CartState = {
  cartId: null,
  cart: null,
  loading: false,
  error: null,
};

// =============================================================================
// Commerce Runtime Context Types
// =============================================================================

export type CommerceRuntimeContextValue = {
  // Site metadata
  siteFamily: string | null;
  commerceProvider: string | null;

  // Catalog data
  regions: MedusaRegion[];
  products: MedusaProduct[];
  collections: MedusaCollection[];
  categories: MedusaCategory[];

  // Current product (if specified)
  currentProduct: MedusaProduct | null;

  // Current category (if category_handle provided)
  currentCategory: MedusaCategory | null;

  // Cart state
  cart: MedusaCart | null;
  cartLoading: boolean;
  cartError: string | null;

  // Cart actions
  createCart: (regionId: string, countryCode?: string, email?: string) => Promise<MedusaCart>;
  getCart: (cartId: string) => Promise<MedusaCart>;
  updateCart: (updates: { email?: string; shipping_address?: MedusaCartAddress; billing_address?: MedusaCartAddress }) => Promise<MedusaCart>;
  addLineItem: (variantId: string, quantity: number, cartIdOverride?: string) => Promise<MedusaCart>;
  updateLineItem: (lineId: string, quantity: number) => Promise<MedusaCart>;
  removeLineItem: (lineId: string) => Promise<void>;

  // Shipping actions
  getShippingOptions: () => Promise<MedusaShippingOption[]>;
  addShippingMethod: (optionId: string) => Promise<MedusaCart>;

  // Payment actions
  getPaymentProviders: (regionId: string) => Promise<MedusaPaymentProvider[]>;
  initializePaymentSession: (providerId: string) => Promise<MedusaPaymentCollection>;
  completeCheckout: () => Promise<SiteCommerceCompleteResponse>;

  // Refresh data
  refreshProducts: (collectionId?: string, categoryId?: string) => Promise<void>;
  refreshCategories: () => Promise<void>;
  refreshCollections: () => Promise<void>;
};

File: mos/backend/app/schemas/commerce.py
from __future__ import annotations

from typing import Any, Optional

from pydantic import AnyUrl, BaseModel


class PublicCheckoutRequest(BaseModel):
    funnelSlug: str
    variantId: Optional[str] = None
    selection: dict[str, Any]
    quantity: int
    successUrl: AnyUrl
    cancelUrl: AnyUrl
    pageId: Optional[str] = None
    visitorId: Optional[str] = None
    sessionId: Optional[str] = None
    utm: Optional[dict[str, Any]] = None


# =============================================================================
# Medusa Store API Response Schemas
# =============================================================================


class MedusaRegion(BaseModel):
    """Medusa region data from Store API."""

    id: str
    name: str
    currency_code: str
    tax_rate: Optional[float] = None
    countries: list[dict[str, Any]] = []


class MedusaProductVariant(BaseModel):
    """Medusa product variant data from Store API."""

    id: str
    title: str
    sku: Optional[str] = None
    barcode: Optional[str] = None
    allow_backorder: bool = False
    manage_inventory: bool = True
    inventory_quantity: Optional[int] = None
    prices: list[dict[str, Any]] = []
    options: dict[str, str] = {}


class MedusaProduct(BaseModel):
    """Medusa product data from Store API."""

    id: str
    title: str
    handle: str
    description: Optional[str] = None
    thumbnail: Optional[str] = None
    status: str = "published"
    variants: list[dict[str, Any]] = []
    options: list[dict[str, Any]] = []
    collection_id: Optional[str] = None
    categories: list[dict[str, Any]] = []


class MedusaCollection(BaseModel):
    """Medusa collection data from Store API."""

    id: str
    title: str
    handle: str
    products: list[dict[str, Any]] = []


class MedusaCategory(BaseModel):
    """Medusa product category data from Store API."""

    id: str
    name: str
    handle: str
    description: Optional[str] = None
    parent_category_id: Optional[str] = None
    category_children: list[dict[str, Any]] = []


class MedusaCartAddress(BaseModel):
    """Medusa cart address data."""

    first_name: Optional[str] = None
    last_name: Optional[str] = None
    address_1: Optional[str] = None
    address_2: Optional[str] = None
    city: Optional[str] = None
    province: Optional[str] = None
    postal_code: Optional[str] = None
    country_code: Optional[str] = None
    phone: Optional[str] = None


class MedusaCartLineItem(BaseModel):
    """Medusa cart line item data."""

    id: str
    cart_id: str
    title: str
    description: Optional[str] = None
    variant_id: str
    quantity: int
    unit_price: int
    subtotal: Optional[int] = None
    tax_total: Optional[int] = None
    total: Optional[int] = None
    variant: Optional[dict[str, Any]] = None


class MedusaCart(BaseModel):
    """Medusa cart data from Store API."""

    id: str
    email: Optional[str] = None
    region_id: str
    shipping_address: Optional[dict[str, Any]] = None
    billing_address: Optional[dict[str, Any]] = None
    items: list[dict[str, Any]] = []
    shipping_methods: list[dict[str, Any]] = []
    subtotal: Optional[int] = None
    tax_total: Optional[int] = None
    shipping_total: Optional[int] = None
    discount_total: Optional[int] = None
    total: Optional[int] = None
    currency_code: str


class MedusaShippingOption(BaseModel):
    """Medusa shipping option data."""

    id: str
    name: str
    price_type: str
    amount: Optional[int] = None
    currency_code: Optional[str] = None
    region_id: str


class MedusaPaymentProvider(BaseModel):
    """Medusa payment provider data."""

    id: str


class MedusaPaymentSession(BaseModel):
    """Medusa payment session data."""

    id: str
    provider_id: str
    status: str
    amount: int
    data: dict[str, Any] = {}


class MedusaPaymentCollection(BaseModel):
    """Medusa payment collection data."""

    id: str
    status: str
    amount: int
    currency_code: str
    payment_sessions: list[dict[str, Any]] = []


# =============================================================================
# Site Commerce API Request/Response Schemas
# =============================================================================


class SiteCommerceCartCreateRequest(BaseModel):
    """Request to create a new cart."""

    region_id: str
    country_code: Optional[str] = None
    email: Optional[str] = None
    shipping_address: Optional[dict[str, Any]] = None
    items: Optional[list[dict[str, Any]]] = None


class SiteCommerceCartUpdateRequest(BaseModel):
    """Request to update a cart."""

    email: Optional[str] = None
    shipping_address: Optional[dict[str, Any]] = None
    billing_address: Optional[dict[str, Any]] = None


class SiteCommerceLineItemAddRequest(BaseModel):
    """Request to add a line item to cart."""

    variant_id: str
    quantity: int = 1


class SiteCommerceLineItemUpdateRequest(BaseModel):
    """Request to update a line item quantity."""

    quantity: int


class SiteCommerceShippingMethodRequest(BaseModel):
    """Request to add a shipping method."""

    cart_id: str
    option_id: str


class SiteCommercePaymentSessionRequest(BaseModel):
    """Request to initialize a payment session."""

    provider_id: str


class SiteCommerceProductQuery(BaseModel):
    """Query parameters for fetching products."""

    product_id: Optional[str] = None
    product_handle: Optional[str] = None
    collection_id: Optional[str] = None
    category_id: Optional[str] = None
    limit: int = 100
    offset: int = 0


class SiteCommerceData(BaseModel):
    """Rich commerce data for site pages."""

    # Site metadata
    siteFamily: Optional[str] = None
    commerceProvider: Optional[str] = None

    # Catalog data
    regions: list[dict[str, Any]] = []
    products: list[dict[str, Any]] = []
    collections: list[dict[str, Any]] = []
    categories: list[dict[str, Any]] = []

    # Current product (if specified)
    currentProduct: Optional[dict[str, Any]] = None

    # Cart data (if cart_id provided)
    cart: Optional[dict[str, Any]] = None

    # Shipping options (if cart_id provided)
    shippingOptions: list[dict[str, Any]] = []

    # Payment providers (if region_id provided)
    paymentProviders: list[dict[str, Any]] = []

    # Default payment provider (if region_id provided)
    defaultPaymentProviderId: Optional[str] = None


class SiteCommerceCartResponse(BaseModel):
    """Response for cart operations."""

    cart: dict[str, Any]


class SiteCommerceShippingOptionsResponse(BaseModel):
    """Response for shipping options."""

    shipping_options: list[dict[str, Any]]


class SiteCommercePaymentProvidersResponse(BaseModel):
    """Response for payment providers."""

    payment_providers: list[dict[str, Any]]
    default_payment_provider_id: Optional[str] = None


class SiteCommercePaymentSessionResponse(BaseModel):
    """Response for payment session initialization."""

    payment_collection: dict[str, Any]


class SiteCommerceCompleteResponse(BaseModel):
    """Response for cart completion."""

    type: str
    order: Optional[dict[str, Any]] = None
    cart: Optional[dict[str, Any]] = None
